### PR TITLE
fix(treesitter): treecursor regression

### LINKED ISSUE #27942: ERROR in finalizer: bad argument #(...) (treesitter_treecursor expected, got no value)
### Problem

I'm not sure if it's related to Neovim. Having fresh (master) nvim-treesitter with enabled highlight and indentation I get

```lua
{ naERROR in finalizer: bad argument #-10004 to 'match' (treesitter_treecursor expected, got no value)
	init.lua [+]                                                                                          stack traceback:                                                                                                        70,5           Bot
		-- INSERT --                                                                                                            [C]: in function 'match'
		...from-src/share/nvim/runtime/lua/vim/treesitter/query.lua:745: in function 'match_preds'
		...from-src/share/nvim/runtime/lua/vim/treesitter/query.lua:955: in function '(for generator)'
		...c/share/nvim/runtime/lua/vim/treesitter/languagetree.lua:855: in function 'f'
		...c/share/nvim/runtime/lua/vim/treesitter/languagetree.lua:190: in function 'tcall'
		...c/share/nvim/runtime/lua/vim/treesitter/languagetree.lua:381: in function '_add_injections'
		...c/share/nvim/runtime/lua/vim/treesitter/languagetree.lua:444: in function 'parse'
		...rc/share/nvim/runtime/lua/vim/treesitter/highlighter.lua:391: in function <...rc/share/nvim/runtime/lua/vim/treesitter/highlighter.lua:386>
		ERROR in finalizer: bad argument #-10004 to 'match' (treesitter_treecursor expected, got no value)
		stack traceback:

		[C]: in function 'match'
		...from-src/share/nvim/runtime/lua/vim/treesitter/query.lua:745: in function 'match_preds'
			...from-src/share/nvim/runtime/lua/vim/treesitter/query.lua:955: in function '(for generator)'
			...c/share/nvim/runtime/lua/vim/treesitter/languagetree.lua:855: in function 'f'
			...c/share/nvim/runtime/lua/vim/treesitter/languagetree.lua:190: in function 'tcall'
			...c/share/nvim/runtime/lua/vim/treesitter/languagetree.lua:381: in function '_add_injections'
			...c/share/nvim/runtime/lua/vim/treesitter/languagetree.lua:444: in function 'parse'
			...rc/share/nvim/runtime/lua/vim/treesitter/highlighter.lua:391: in function <...rc/share/nvim/runtime/lua/vim/treesitter/highlighter.lua:386>                                                                                                                                                                                                                           6
```

when make some edits.


### Steps to reproduce

init.lua:
```lua
local thisInitFile = debug.getinfo(1).source:match('@?(.*)')
local configDir = vim.fs.dirname(thisInitFile)

vim.env['XDG_CONFIG_HOME'] = configDir
vim.env['XDG_DATA_HOME'] = vim.fs.joinpath(configDir, '.xdg', 'data')
vim.env['XDG_STATE_HOME'] = vim.fs.joinpath(configDir, '.xdg', 'state')
vim.env['XDG_CACHE_HOME'] = vim.fs.joinpath(configDir, '.xdg', 'cache')
local stdPathConfig = vim.fn.stdpath('config')

vim.opt.runtimepath:prepend(stdPathConfig)
vim.opt.packpath:prepend(stdPathConfig)

local function gitClone(url, installPath, branch)
  if vim.fn.isdirectory(installPath) ~= 0 then
    return
  end

  local command = {'git', 'clone', '--', url, installPath}
  if branch then
    table.insert(command, 3, '--branch')
    table.insert(command, 4, branch)
  end
  local sysObj = vim.system(command, {}):wait()
  if sysObj.code ~= 0 then
    error(sysObj.stderr)
  end
  vim.notify(sysObj.stdout)
  vim.notify(sysObj.stderr, vim.log.levels.WARN)
end

local pluginsPath = 'nvim/pack/plugins/opt'
vim.fn.mkdir(pluginsPath, 'p')
pluginsPath = vim.uv.fs_realpath(pluginsPath)

--- @type table<string, {url:string, branch: string?}>
local plugins = {
  ['nvim-treesitter'] = {url = 'https://github.com/nvim-treesitter/nvim-treesitter'},
}

for name, repo in pairs(plugins) do
  local installPath = vim.fs.joinpath(pluginsPath, name)
  gitClone(repo.url, installPath, repo.branch)
  -- vim.opt.runtimepath:append(installPath)
  vim.cmd.packadd({args = {name}, bang = true})
end

require 'nvim-treesitter.configs'.setup {
  ensure_installed = {
  },
  highlight = {
    enable = true,
  },
  incremental_selection = {
    enable = true,
    keymaps = tsKeymaps,
  },
  indent = {
    enable = true,
  },
}
```

- Run `nvim-src-clean -u init.lua init.lua`
- type `Go` (start Insert mode at the end of file)
- insert empty `{}` pair and then inside start typing `name = ''` - for me it's enough to reproduce this error. 

### Expected behavior

No error

### Neovim version (nvim -v)

v0.10.0-dev-2656+g37db3d97e

### Vim (not Nvim) behaves the same?

-

### Operating system/version

Debian Sid

### Terminal name/version

alacritty 0.13.1, tmux 3.4-2

### $TERM environment variable

tmux-256color

### Installation

from repo

### PR DESCRIPTION
**Pull request by lewis6991**
Fixes #27942

### CODE REVIEW

**Comment by clason** [comment]
This fixes the issue with `format-queries` 👍

**Comment by lewis6991** [comment]
I found a couple typos that should address at least one error.

File: src/nvim/lua/treesitter.c [hunk 1]
```diff
@@ -1032,7 +1037,10 @@ static int node_named_descendant_for_range(lua_State *L)
 
 static int node_next_child(lua_State *L)
 {
-  TSTreeCursor *cursor = treecursor_check(L, lua_upvalueindex(1));
```
Comment (lewis6991) [inline src/nvim/lua/treesitter.c:1035]: I'm not sure why this failed. The cursor should always be an upvalue.

File: src/nvim/lua/treesitter.c [hunk 2]
```diff
@@ -715,9 +718,10 @@ static struct luaL_Reg treecursor_meta[] = {
 
 static TSTreeCursor *treecursor_check(lua_State *L, int index)
 {
-  TSTreeCursor *ud = luaL_checkudata(L, lua_upvalueindex(1), TS_META_TREECURSOR);
+  TSTreeCursor *ud = luaL_checkudata(L, index, TS_META_TREECURSOR);
```
Comment (lewis6991) [inline src/nvim/lua/treesitter.c:721]: typo fix

File: src/nvim/lua/treesitter.c [hunk 3]
```diff
@@ -1310,9 +1319,10 @@ static void push_querymatch(lua_State *L, TSQueryMatch *match, int uindex)
 
 static TSQueryMatch *querymatch_check(lua_State *L, int index)
 {
-  TSQueryMatch *ud = luaL_checkudata(L, 1, TS_META_QUERYMATCH);
+  TSQueryMatch *ud = luaL_checkudata(L, index, TS_META_QUERYMATCH);
```
Comment (lewis6991) [inline src/nvim/lua/treesitter.c:1322]: typo fix

File: src/nvim/lua/treesitter.c (hunk 1 above)
Comment (clason) [reply src/nvim/lua/treesitter.c:1035]: It doesn't; must have been the `gc` without the early return.

File: src/nvim/lua/treesitter.c (hunk 3 above)
Comment (clason) [reply src/nvim/lua/treesitter.c:1317]: This fix is enough to prevent the `format-queries` error.

**Comment by przepompownia** [comment]
With this change I get no errors both using the repro from #27942 and `format-queries`.

**Pull request merged**